kernel/util/mlog-base/util/Logger.hh

```diff
@@ -51,51 +51,79 @@ namespace mlog {
+    if (this->isActive(MSG::VERBOSITY)) write<MSG, ARGS...>(std::forward<ARGS>(args)...);
+  }
 
-    template<class MSG, typename... ARGS>
-    void write(ARGS&&... args) const NOINLINE;
+  template<class MSG, typename... ARGS>
+  void write(ARGS&&... args) const NOINLINE;
 
-    void flush() { sink->flush(); }
+  void flush() { sink->flush(); }
 
-    // syntactic sugar
-    template<typename... ARGS>
-    void detail(ARGS&&... args) { log<TextDetail>(std::forward<ARGS>(args)...); }
+  // syntactic sugar
+  template<typename... ARGS>
+  void detail(ARGS&&... args) { log<TextDetail>(std::forward<ARGS>(args)...); }
 
-    template<typename... ARGS>
-    void info(ARGS&&... args) { log<TextInfo>(std::forward<ARGS>(args)...); }
+  template<typename... ARGS>
+  void info(ARGS&&... args) { log<TextInfo>(std::forward<ARGS>(args)...); }
 
-    template<typename... ARGS>
-    void warn(ARGS&&... args) { log<TextWarning>(std::forward<ARGS>(args)...); }
+  template<typename... ARGS>
+  void warn(ARGS&&... args) { log<TextWarning>(std::forward<ARGS>(args)...); }
 
-    template<typename... ARGS>
-    void error(ARGS&&... args) { log<TextError>(std::forward<ARGS>(args)...); }
+  template<typename... ARGS>
+  void error(ARGS&&... args) { log<TextError>(std::forward<ARGS>(args)...); }
 
-    template<typename... ARGS>
-    void csv(ARGS&&... args) { log<TextCSV>(std::forward<ARGS>(args)...); }
+  template<typename... ARGS>
+  void csv(ARGS&&... args) { log<TextCSV>(std::forward<ARGS>(args)...); }
 
-  protected:
-    char const* name;
-  };
+protected:
+  char const* name;
+};
 
-  template<class Filter>
-  template<class MSG, typename... ARGS>
-  void Logger<Filter>::write(ARGS&&... args) const {
-    MSG msg(name, std::forward<ARGS>(args)...);
-    sink->write(msg.getData(), msg.getSize());
+template<class Filter>
+template<class MSG, typename... ARGS>
+void Logger<Filter>::write(ARGS&&... args) const {
+  MSG msg(name, std::forward<ARGS>(args)...);
+  sink->write(msg.getData(), msg.getSize());
+}
+
+static Logger<FilterAny> testLog("Test");
+inline static void test_log(const char *file_line, const char *msg, const char *expr) {
+  if (testLog.isActive(mlog::TextError::VERBOSITY)) {
+    testLog.write<mlog::TextError>(file_line, msg, expr);
   }
+}
+
+#define MLOG_TEST_OP(logger, op, val, expected) \
+  if (!((val) op (expected))) MLOG_ERROR(logger, "Failed Test:", \
+  TOSTRING(val), TOSTRING(op), TOSTRING(expected))
+#define MLOG_TEST(logger, expr) if (!(expr)) MLOG_ERROR(logger, "Failed Test:", TOSTRING(expr))
+
+#define MLOG_TEST_EQ(logger, val, expected) MLOG_TEST_OP(logger, ==, val, expected)
+#define MLOG_TEST_NEQ(logger, expr, val) MLOG_TEST_OP(logger, !=, expr, val)
+#define MLOG_TEST_TRUE(logger, expr) MLOG_TEST_OP(logger, ==, expr, true)
+#define MLOG_TEST_FALSE(logger, expr) MLOG_TEST_OP(logger, !=, expr, true)
+
+#define TEST(expr) if (!(expr)) mlog::test_log("[" __FILE__ ":" TOSTRING(__LINE__) "]", "Failed Test:", TOSTRING(expr))
```

Comment:
An opportunity for simplification: These macros should just call MLOG_TEST_* with the default logging channel `testLog`. This would also ensure, that the output formatting is the same for both channels! For the file,line pair you can add a helper structure in TextMsgHelpers. It just contains both string pointers and `operator<<`. Or reuse the PATH macro from the beginning of Logger.hh.